ceramic-ahash/src/ahash.rs

```diff
@@ -0,0 +1,210 @@
+use ::serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
+use multihash::{Hasher, Sha2_256};
+use std::fmt::{self, Debug, Display, Formatter};
+
+use crate::Hash;
+
+#[derive(Default, PartialEq, Clone, Copy, Deserialize)]
+pub struct AHash(#[serde(deserialize_with = "deserialize_from_bytes")] pub [u32; 8]);
+
+impl std::ops::Add for AHash {
+    type Output = Self;
+    fn add(self, rhs: Self) -> Self::Output {
+        AHash([
+            self.0[0].wrapping_add(rhs.0[0]),
+            self.0[1].wrapping_add(rhs.0[1]),
+            self.0[2].wrapping_add(rhs.0[2]),
+            self.0[3].wrapping_add(rhs.0[3]),
+            self.0[4].wrapping_add(rhs.0[4]),
+            self.0[5].wrapping_add(rhs.0[5]),
+            self.0[6].wrapping_add(rhs.0[6]),
+            self.0[7].wrapping_add(rhs.0[7]),
+        ])
+    }
+}
+
+impl Serialize for AHash {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_bytes(&self.to_bytes())
+    }
+}
+
+fn deserialize_from_bytes<'de, D>(deserializer: D) -> Result<[u32; 8], D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let b: &serde_bytes::Bytes = Deserialize::deserialize(deserializer)?;
+    Ok(AHash::from_bytes(
+        b.to_vec()
+            .as_slice()
+            .try_into()
+            .map_err(|_| D::Error::invalid_length(b.len(), &"hash must be len 32"))?,
+    )
+    .0)
+    // todo return error on errors
+}
+
+impl crate::recon::Hash for AHash {
+    fn is_zero(&self) -> bool {
+        self == &Self::default()
+    }
+
+    fn clear(&mut self) {
+        self.0[0] = 0;
+        self.0[1] = 0;
+        self.0[2] = 0;
+        self.0[3] = 0;
+        self.0[4] = 0;
+        self.0[5] = 0;
+        self.0[6] = 0;
+        self.0[7] = 0;
+    }
+    fn push(&mut self, key: (&String, &AHash)) {
```

Comment:
What is the `key.0` (the string) for? Does not look to be used in this implementation. Do other implementations need the key string?

Reply:
MemoryAHash.push((&String, &AHash)) needs to remember the string and not just the hash